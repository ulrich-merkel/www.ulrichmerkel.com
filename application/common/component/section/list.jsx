/* eslint-disable immutable/no-mutation */
/**
 * Es6 module for React Component.
 * Section components combine modules and elements
 * to create a page section which could be easily
 * used in page components.
 *
 * @file
 * @module
 * @flow weak
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2016
 * @version 0.0.4
 *
 * @requires react
 * @requires prop-types
 * @requires common/component/section/common/grid-spaced
 * @requires common/component/module/article
 * @requires common/component/module/list
 *
 * @changelog
 * - 0.0.1 Basic functions and structure
 */
import * as React from 'react';
import PropTypes from 'prop-types';

import SectionCommonGridSpaced from './common/grid-spaced';
import ModuleArticle from '../module/article';
import ModuleList from '../module/list';

/**
 * Function representing a component to return a single react child element.
 *
 * @function
 * @param {object} [props] - The current component props
 * @param {Array|string} [props.children] - The component dom node childs - usually an array of components, if there is only a single child it's a string
 * @param {object} [props.content={}] - The content config input
 * @param {boolean} [props.isDialog=false] - Whether the component text should be displayed in a dialog or not
 * @param {boolean} [props.isMain=false] - Whether the component text should be displayed as main article or not
 * @returns {ReactElement} React component markup
 */
function SectionList(props) {
    const { children, content, isDialog, isMain, ...moduleProps } = props;

    return (
        <SectionCommonGridSpaced>
            <ModuleArticle {...{ content, isMain, isDialog }}>
                <ModuleList {...{ content }} {...moduleProps}>
                    {children}
                </ModuleList>
            </ModuleArticle>
        </SectionCommonGridSpaced>
    );
}

/**
 * Validate props via React.PropTypes helpers.
 *
 * @static
 * @type {object}
 */
SectionList.propTypes = {
    children: PropTypes.node, // eslint-disable-line react/require-default-props
    content: PropTypes.objectOf(
        PropTypes.oneOfType([
            PropTypes.string,
            PropTypes.number,
            PropTypes.array,
            PropTypes.object
        ])
    ),
    isDialog: PropTypes.bool,
    isMain: PropTypes.bool
};

/**
 * Set defaults if props aren't available.
 *
 * @static
 * @type {object}
 */
SectionList.defaultProps = {
    content: {},
    isDialog: false,
    isMain: false
};

export default SectionList;
