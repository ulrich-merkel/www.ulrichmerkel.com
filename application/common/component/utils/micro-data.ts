/**
 * Es6 module for creating mirco data attributes.
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2021
 */
import { isValidString } from '../../utils/string';

type ItemTypeAttributes = {
    itemScope: boolean;
    itemType: string;
};

/**
 * Get component properties for itemProp and itemType.
 *
 * @param {boolean} itemType - The itemType property if available
 * @returns {object|null} The component props for spreading
 */
export function getItemTypeAttributes(
    itemType?: string
): ItemTypeAttributes | null {
    return isValidString(itemType)
        ? {
              itemScope: true,
              itemType
          }
        : null;
}
