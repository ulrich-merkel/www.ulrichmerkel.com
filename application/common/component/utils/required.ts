/**
 * Es6 module for creating form input required attributes.
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2021
 */
import { isBoolean } from 'lodash';

type RequiredAttributes = {
    'aria-required': boolean;
    required: boolean;
};

/**
 * Get component properties for required form input elements.
 *
 * @param {boolean} required - Whether the input element is required or not
 * @returns {object|null} The component props for spreading
 */
export function getRequiredAttributes(
    required?: boolean
): RequiredAttributes | null {
    return isBoolean(required) && required
        ? {
              'aria-required': true,
              required: true
          }
        : null;
}
