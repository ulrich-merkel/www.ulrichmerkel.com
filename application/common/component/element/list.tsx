/**
 * Rendering a small html tag.
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2021
 */
import { default as React, FunctionComponent, ReactNode } from 'react';
import classnames from 'classnames';
import { getItemTypeAttributes } from '../utils/micro-data';

type Props = {
    children?: ReactNode;
    className?: string;
    htmlElement?: keyof JSX.IntrinsicElements;
    itemProp?: string;
    itemType?: string;
    role?: string;
};

/**
 * Function representing a html list (ul/ol) element.
 *
 * @function
 * @param {object} [props] - The current component props
 * @returns {ReactElement} React component markup
 */
export const List: FunctionComponent<Props> = (props) => {
    const {
        children,
        className,
        htmlElement: HtmlElement = 'ul',
        itemProp,
        itemType,
        role
    } = props;

    const composedClassName = classnames('c-list', className);
    const itemTypeAttributes = getItemTypeAttributes(itemType);

    return (
        <HtmlElement
            className={composedClassName}
            {...{ itemProp, role }}
            {...itemTypeAttributes}
        >{children}</HtmlElement>
    );
};
