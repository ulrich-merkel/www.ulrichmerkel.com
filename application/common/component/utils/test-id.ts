/**
 * Es6 module for creating test id related attributes
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2021
 */
import { isValidString } from '../../utils/string';

type TestIdAttributes = {
    'data-testid': string;
};

/**
 * Get component properties for testing.
 *
 * @param {string} testId - The test id to be added
 * @returns {object|null} The component props for spreading
 */
export function getTestIdAttributes(testId?: string): TestIdAttributes | null {
    return isValidString(testId)
        ? {
              'data-testid': testId
          }
        : null;
}
