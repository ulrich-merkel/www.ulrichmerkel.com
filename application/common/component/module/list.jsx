/* eslint-disable react/no-danger, immutable/no-mutation */
/**
 * Es6 module for React Component.
 * Component module React classes combine elements to
 * bigger parts of the page.
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2021
 */
import * as React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import shortid from 'shortid';

import { Headline } from '../element/headline';

/**
 * Function representing a component to return a single react child element.
 *
 * @param {object} [props] - The current component props
 * @returns {ReactElement} React component markup
 */
export function ModuleList(props) {
    const { componentType, className, itemType, content, children } = props;

    const ComponentType = componentType;
    const componentClassName = classnames('m-list--broadcast', className);

    return (
        <ComponentType
            className={componentClassName}
            itemScope
            itemType={itemType}
            role="list"
        >
            {content.text &&
                content.text.map((entry) => {
                    return (
                        <li className="m-list__item" key={shortid.generate()}>
                            <Headline
                                className="m-list__alt-headline"
                                isCentered={false}
                                htmlElement="h3"
                            >
                                {entry.headline}
                            </Headline>
                            <ul className="m-list">
                                {entry.list &&
                                    entry.list.map((job) => {
                                        const text = `${job.name}, ${job.job} - ${job.place}`;
                                        return (
                                            <li
                                                key={shortid.generate()}
                                                className="m-list__item"
                                                itemProp="itemListElement"
                                                dangerouslySetInnerHTML={{
                                                    __html: text
                                                }}
                                            />
                                        );
                                    })}
                            </ul>
                        </li>
                    );
                })}
            {children}
        </ComponentType>
    );
}

/**
 * Validate props via React.PropTypes helpers.
 *
 * @static
 * @type {object}
 * @property {string} [componentType='ul'] - The component element type used for React.createElement
 * @property {string} [className] - The component css class names - will be merged into component default classNames
 * @property {string} [itemType='http://schema.org/ItemList'] - The schema.org itemtype url attribute
 * @property {Array|string} [children] - The component dom node childs - usally an array of components, if there is only a single child it's a string
 * @property {object} [content={}] - The component translation config
 */
ModuleList.propTypes = {
    componentType: PropTypes.string,
    className: PropTypes.string, // eslint-disable-line react/require-default-props
    itemType: PropTypes.string,
    children: PropTypes.node, // eslint-disable-line react/require-default-props
    content: PropTypes.objectOf(
        PropTypes.oneOfType([
            PropTypes.string,
            PropTypes.number,
            PropTypes.array,
            PropTypes.object
        ])
    )
};

/**
 * Set defaults if props aren't available.
 *
 * @static
 * @type {object}
 * @see ModuleList.propTypes
 */
ModuleList.defaultProps = {
    componentType: 'ul',
    itemType: 'http://schema.org/ItemList',
    content: {}
};
