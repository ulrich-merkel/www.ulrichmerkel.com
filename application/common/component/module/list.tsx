/**
 * Es6 module for React Component.
 * Component module React classes combine elements to
 * bigger parts of the page.
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2021
 */
import { default as React, FunctionComponent, ReactNode } from 'react';
import classnames from 'classnames';
import shortid from 'shortid';

import { Headline } from '../element/headline';
import { ListItem } from '../element/list-item';
import { getItemTypeAttributes } from '../utils/micro-data';
import { isValidArray } from '../../utils/array';
import { List } from '../element/list';

type Props = {
    children?: ReactNode;
    className?: string;
    content?: Record<string, unknown>;
    htmlElement?: keyof JSX.IntrinsicElements;
    itemType?: string;
    role?: string;
};

/**
 * Function representing a component to return a single react child element.
 *
 * @function
 * @param {object} [props] - The current component props
 * @returns {ReactElement} React component markup
 */
export const ModuleList: FunctionComponent<Props> = (props) => {
    const {
        children,
        className,
        content,
        htmlElement: HtmlElement = 'ul',
        itemType = 'http://schema.org/ItemList',
        role = 'list'
    } = props;

    const componentClassName = classnames('m-list--broadcast', className);
    const itemTypeAttributes = getItemTypeAttributes(itemType);

    if (!isValidArray(content?.text)) {
        return null;
    }

    return (
        <HtmlElement
            className={componentClassName}
            {...itemTypeAttributes}
            {...{ role }}
        >
            {content.text.map(function fnMap(entry) {
                return (
                    <ListItem className="m-list__item" key={shortid.generate()}>
                        <Headline
                            className="m-list__alt-headline"
                            isCentered={false}
                            htmlElement="h3"
                        >
                            {entry.headline}
                        </Headline>
                        <List className="m-list">
                            {isValidArray(entry.list) &&
                                entry.list.map(function fnMapJob(job) {
                                    const text = `${job.name}, ${job.job} - ${job.place}`;
                                    return (
                                        <ListItem
                                            key={shortid.generate()}
                                            className="m-list__item"
                                            itemProp="itemListElement"
                                            dangerouslySetInnerHTML={{
                                                __html: text
                                            }}
                                        />
                                    );
                                })}
                        </List>
                    </ListItem>
                );
            })}
            {children}
        </HtmlElement>
    );
};
