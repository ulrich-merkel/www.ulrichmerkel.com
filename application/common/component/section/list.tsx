/**
 * Es6 module for React Component.
 * Section components combine modules and elements
 * to create a page section which could be easily
 * used in page components.
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2021
 */
import { default as React, FunctionComponent, ReactNode } from 'react';

import { SectionCommonGridSpaced } from './common/grid-spaced';
import { ModuleArticle } from '../module/article';
import { ModuleList } from '../module/list';

type Props = {
    children?: ReactNode;
    content: Record<string, unknown>;
    isDialog?: boolean;
    isMain?: boolean;
};

/**
 * Function representing a component to return a single react child element.
 *
 * @function
 * @param {object} [props] - The current component props
 * @param {Array|string} [props.children] - The component dom node childs - usually an array of components, if there is only a single child it's a string
 * @param {object} [props.content={}] - The content config input
 * @param {boolean} [props.isDialog=false] - Whether the component text should be displayed in a dialog or not
 * @param {boolean} [props.isMain=false] - Whether the component text should be displayed as main article or not
 * @returns {ReactElement} React component markup
 */
export const SectionList: FunctionComponent<Props> = (props) => {
    const { children, content, isDialog, isMain, ...otherProps } = props;

    return (
        <SectionCommonGridSpaced>
            <ModuleArticle {...{ content, isMain, isDialog }}>
                <ModuleList {...{ content }} {...otherProps}>
                    {children}
                </ModuleList>
            </ModuleArticle>
        </SectionCommonGridSpaced>
    );
};
