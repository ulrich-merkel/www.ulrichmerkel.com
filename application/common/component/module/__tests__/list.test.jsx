import * as React from 'react';
import { render } from '../../../../__tests__/utils/test-utils';
import { ModuleList } from '../list';

describe('ModuleList', function fnDescribe() {
    const defaultProps = {
        content: {
            text: [
                {
                    headline: '2012',
                    list: [
                        {
                            name:
                                "<a href='http://www.diamondleague.com/' target='_blank' title='Diamant League website'>Diamant League</a>, track and field athletics",
                            job: 'intl-broadcast-virtual-graphics',
                            place: 'USA (New York, Eugene)'
                        }
                    ]
                },
                {
                    headline: '2011',
                    list: [
                        {
                            name:
                                "<a href='http://www.bundesliga.de/' target='_blank' title='Bundesliga website'>German Soccer League</a>",
                            job: 'intl-broadcast-virtual-graphics',
                            place: 'Germany'
                        },
                        {
                            name:
                                "<a href='http://www.mdr.de/' target='_blank' title='MDR website'>Mitteldeutscher Rundfunk Leipzig</a>",
                            job: 'intl-broadcast-tv-graphics',
                            place: 'Germany'
                        }
                    ]
                }
            ]
        },
        className: 'module-list',
        componentType: 'ol'
    };

    it('should render correctly', function fnIt() {
        const { asFragment } = render(
            <ModuleList {...defaultProps}>Module list children</ModuleList>
        );
        expect(asFragment()).toMatchSnapshot();
    });
});
