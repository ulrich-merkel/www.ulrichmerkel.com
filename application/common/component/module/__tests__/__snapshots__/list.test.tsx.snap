// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModuleList should render correctly 1`] = `
<DocumentFragment>
  <ol
    class="m-list--broadcast module-list"
    itemscope=""
    itemtype="http://schema.org/ItemList"
    role="list"
  >
    <li
      class="c-list-item m-list__item"
      itemprop="itemListElement"
    >
      <h3
        class="c-type--h3 m-list__alt-headline"
      >
        2012
      </h3>
      <ul
        class="c-list m-list"
        itemscope=""
        itemtype="https://schema.org/ItemList"
        role="list"
      >
        <li
          class="c-list-item m-list__item"
          itemprop="itemListElement"
        >
          <a
            href="http://www.diamondleague.com/"
            target="_blank"
            title="Diamant League website"
          >
            Diamant League
          </a>
          , track and field athletics, intl-broadcast-virtual-graphics - USA (New York, Eugene)
        </li>
      </ul>
    </li>
    <li
      class="c-list-item m-list__item"
      itemprop="itemListElement"
    >
      <h3
        class="c-type--h3 m-list__alt-headline"
      >
        2011
      </h3>
      <ul
        class="c-list m-list"
        itemscope=""
        itemtype="https://schema.org/ItemList"
        role="list"
      >
        <li
          class="c-list-item m-list__item"
          itemprop="itemListElement"
        >
          <a
            href="http://www.bundesliga.de/"
            target="_blank"
            title="Bundesliga website"
          >
            German Soccer League
          </a>
          , intl-broadcast-virtual-graphics - Germany
        </li>
        <li
          class="c-list-item m-list__item"
          itemprop="itemListElement"
        >
          <a
            href="http://www.mdr.de/"
            target="_blank"
            title="MDR website"
          >
            Mitteldeutscher Rundfunk Leipzig
          </a>
          , intl-broadcast-tv-graphics - Germany
        </li>
      </ul>
    </li>
    Module list children
  </ol>
</DocumentFragment>
`;
